include/gsl/span

```diff
@@ -21,13 +21,27 @@
 #include <gsl/gsl_byte>   // for byte
 #include <gsl/gsl_util>   // for narrow_cast, narrow
 
-#include <algorithm> // for lexicographical_compare
```

Comment:
Looks like this PR is addressing two sets of changes:
1. Moving non-standard free functions outside of the span header into its own separate header.
2. Changing the span implementation itself by forward declaring std::array.

Unless there's a technical justification, could you please create separate PRs for the two sets of changes?

Reply:
I saw the elimination of the span header as a necessary aspect of cleaning the span header itself. To me, it made sense to have the efforts be combined into a single PR, however you're right that they're not technically intertwined. So I've reverted the array header removal and have set that as #849

Reply:
@sunnychatterjee I've abandoned the array removal option for now. I need to investigate the UB associated with the forward decl a bit more before committing to one solution or another.